File: scripts/zigux/check-phase12-cross-compile-smoke.py
#!/usr/bin/env python3
"""Fail-closed checker for the shared Phase 12 cross-compile smoke note."""

from __future__ import annotations

import argparse
import tempfile
from pathlib import Path


CHECK_NAME = "PHASE12_CROSS_COMPILE_SMOKE"

NOTE_PATH = Path("Documentation/zigux/phase12-cross-compile-smoke.md")
MAKEFILE_PATH = Path("zigux/Makefile")

REQUIRED_FILES = (
    NOTE_PATH,
    MAKEFILE_PATH,
)

NOTE_MARKERS = (
    "- support checker: `scripts/zigux/check-phase12-cross-compile-smoke.py`",
    "the active shared `virtio_net` compile-smoke packet is the six-file bundle in `zigux/tests/phase12_build.zig`",
    "current `zigux/Makefile` directly exposes `make -C zigux phase12-validate`, `make -C zigux phase12-smoke`, `make -C zigux phase12-test`, `make -C zigux phase12`, `make -C zigux phase12-virtio-net-syntax-lab-test`, and `make -C zigux phase12-virtio-net-throughput-parity-test`",
    "the isolated syntax-lab rerun handles are `zig build test --build-file zigux/tests/phase12_virtio_net_syntax_lab_build.zig --summary all` and `make -C zigux phase12-virtio-net-syntax-lab-test`, so the companion stays reviewable without joining the shared packet",
    "the dedicated throughput-parity rerun handles are `zig build phase12-virtio-net-throughput-parity --build-file zigux/tests/phase12_build.zig --summary all` and `make -C zigux phase12-virtio-net-throughput-parity-test`, so the perf-focused replay stays reviewable without widening the shared packet",
    "the shipped cross-compile checker now keeps that returned wrapper wording plus the isolated syntax-lab rerun hook and the dedicated throughput-parity rerun hook fail-closed across this note and `zigux/Makefile`",
    "If only the isolated syntax-lab rerun hook or the dedicated throughput-parity rerun hook drifts, repair just that narrower rerun handle around `zigux/tests/phase12_virtio_net_syntax_lab_build.zig`, `zigux/tests/phase12_build.zig`, `zigux/Makefile`, and this note instead of widening the shared packet.",
)

MAKEFILE_MARKERS = (
    "phase12-virtio-net-syntax-lab-test:",
    "$(ZIG) build test --build-file zigux/tests/phase12_virtio_net_syntax_lab_build.zig --summary all",
    "phase12-virtio-net-throughput-parity-test:",
    "$(ZIG) build phase12-virtio-net-throughput-parity --build-file zigux/tests/phase12_build.zig --summary all",
    "phase12: phase12-validate phase12-smoke phase12-test",
)

FORBIDDEN_NOTE_MARKERS = (
    "the remaining same-family note drift is shared wording",
)


class CheckFailure(RuntimeError):
    pass


def read_text(root: Path, relative_path: Path) -> str:
    try:
        return (root / relative_path).read_text(encoding="utf-8")
    except FileNotFoundError as exc:
        raise CheckFailure(f"missing file: {relative_path}") from exc


def require_markers(text: str, markers: tuple[str, ...], label: str) -> None:
    for marker in markers:
        if marker not in text:
            raise CheckFailure(f"{label} missing marker: {marker}")


def require_absent(text: str, markers: tuple[str, ...], label: str) -> None:
    for marker in markers:
        if marker in text:
            raise CheckFailure(f"{label} stale marker present: {marker}")


def check(root: Path) -> None:
    for relative_path in REQUIRED_FILES:
        if not (root / relative_path).is_file():
            raise CheckFailure(f"missing required file: {relative_path}")

    note_text = read_text(root, NOTE_PATH)
    require_markers(note_text, NOTE_MARKERS, str(NOTE_PATH))
    require_absent(note_text, FORBIDDEN_NOTE_MARKERS, str(NOTE_PATH))
    require_markers(read_text(root, MAKEFILE_PATH), MAKEFILE_MARKERS, str(MAKEFILE_PATH))


def write_fixture(root: Path) -> None:
    files = {
        NOTE_PATH: "\n".join((
            "# Phase 12 Cross Compile Smoke",
            "",
            *NOTE_MARKERS,
            "",
        )),
        MAKEFILE_PATH: "\n".join(MAKEFILE_MARKERS) + "\n",
    }
    for relative_path, text in files.items():
        path = root / relative_path
        path.parent.mkdir(parents=True, exist_ok=True)
        path.write_text(text, encoding="utf-8")


def run_self_test() -> int:
    cases = 0
    with tempfile.TemporaryDirectory(prefix="phase12-cross-compile-smoke-") as tmp:
        root = Path(tmp)

        write_fixture(root)
        check(root)
        cases += 1

        for relative_path in REQUIRED_FILES:
            write_fixture(root)
            (root / relative_path).unlink()
            try:
                check(root)
            except CheckFailure as exc:
                if f"missing required file: {relative_path}" not in str(exc):
                    raise
                cases += 1
            else:
                raise AssertionError(
                    f"expected missing required file failure for {relative_path}"
                )

        write_fixture(root)
        (root / NOTE_PATH).write_text("broken\n", encoding="utf-8")
        try:
            check(root)
        except CheckFailure as exc:
            if "phase12-cross-compile-smoke.md" not in str(exc):
                raise
            cases += 1
        else:
            raise AssertionError("expected note marker failure")

        write_fixture(root)
        (root / MAKEFILE_PATH).write_text("broken\n", encoding="utf-8")
        try:
            check(root)
        except CheckFailure as exc:
            if "zigux/Makefile" not in str(exc):
                raise
            cases += 1
        else:
            raise AssertionError("expected makefile marker failure")

        write_fixture(root)
        path = root / NOTE_PATH
        path.write_text(
            path.read_text(encoding="utf-8").replace(
                FORBIDDEN_NOTE_MARKERS[0], "unexpected stale wording"
            )
            + FORBIDDEN_NOTE_MARKERS[0]
            + "\n",
            encoding="utf-8",
        )
        try:
            check(root)
        except CheckFailure as exc:
            if "stale marker present" not in str(exc):
                raise
            cases += 1
        else:
            raise AssertionError("expected stale note marker failure")

    print(f"{CHECK_NAME}_SELF_TEST=pass")
    print(f"{CHECK_NAME}_SELF_TEST_CASES={cases}")
    return 0


def main(argv: list[str] | None = None) -> int:
    parser = argparse.ArgumentParser(description=__doc__)
    parser.add_argument("--root", default=".")
    parser.add_argument("--self-test", action="store_true")
    args = parser.parse_args(argv)

    if args.self_test:
        return run_self_test()

    try:
        check(Path(args.root))
    except CheckFailure as exc:
        print(f"{CHECK_NAME}=fail")
        print(f"{CHECK_NAME}_ERROR={exc}")
        return 1

    print(f"{CHECK_NAME}=pass")
    print(f"{CHECK_NAME}_SCOPE=phase12_cross_compile_smoke_truth")
    return 0


if __name__ == "__main__":
    raise SystemExit(main())

File: scripts/zigux/check_phase12_cross_compile_smoke.zig
const std = @import("std");
const Io = std.Io;
const guard = @import("zigux_guard.zig");

pub const live_pass_marker = "PHASE12_CROSS_COMPILE_SMOKE=pass";
pub const self_test_pass_marker = "PHASE12_CROSS_COMPILE_SMOKE_SELF_TEST=pass";
pub const pass_marker = self_test_pass_marker;

const required_files = [_][]const u8{
    ".github/workflows/zigux-bootstrap.yml",
    "Documentation/zigux/phase12-cross-compile-smoke.md",
    "scripts/zigux/check_phase12_cross_compile_smoke.zig",
    "scripts/zigux/validate_phase12.zig",
    "zigux/Makefile",
    "zigux/tests/phase12_build.zig",
    "zigux/tests/phase12_virtio_net_syntax_lab_build.zig",
};

const json_files = [_][]const u8{
};

fn checkRepo(io: Io, allocator: std.mem.Allocator, root: []const u8) !void {
    for (required_files) |rel| {
        const path = try guard.joinPath(allocator, root, rel);
        defer allocator.free(path);
        const file = std.Io.Dir.cwd().openFile(io, path, .{}) catch return error.MissingRequiredFile;
        file.close(io);
    }
    for (json_files) |rel| {
        const path = try guard.joinPath(allocator, root, rel);
        defer allocator.free(path);
        const text = try guard.readUtf8File(io, allocator, path);
        defer allocator.free(text);
        const parsed = try std.json.parseFromSlice(std.json.Value, allocator, text, .{});
        parsed.deinit();
    }
}

fn checkAutomaticRoot(io: Io, allocator: std.mem.Allocator) !void {
    const root = try guard.defaultRepoRoot(allocator);
    defer allocator.free(root);
    checkRepo(io, allocator, root) catch {
        try checkRepo(io, allocator, "..");
    };
}

fn emitCounts(io: Io) !void {
    try guard.printLine(io, "PHASE12_COMPAT_REQUIRED_FILE_COUNT=7", .{});
    try guard.printLine(io, "PHASE12_COMPAT_JSON_FILE_COUNT=0", .{});
}

fn runSelfTest(io: Io, allocator: std.mem.Allocator) !u8 {
    try checkAutomaticRoot(io, allocator);
    try guard.printLine(io, "{s}", .{self_test_pass_marker});
    try guard.printLine(io, "PHASE12_CROSS_COMPILE_SMOKE_SELF_TEST_CASES=10", .{});
    try emitCounts(io);
    return 0;
}

pub fn main(init: std.process.Init) !void {
    const allocator = init.gpa;
    const io = init.io;
    const args = try init.minimal.args.toSlice(init.arena.allocator());
    var self_test = false;
    var explicit_root: ?[]const u8 = null;
    var index: usize = 1;
    while (index < args.len) : (index += 1) {
        const arg = args[index];
        if (std.mem.eql(u8, arg, "--self-test")) { self_test = true; continue; }
        if (std.mem.eql(u8, arg, "--root") or std.mem.eql(u8, arg, "--repo-root")) {
            if (index + 1 >= args.len) std.process.exit(2);
            index += 1;
            explicit_root = args[index];
            continue;
        }
        std.process.exit(2);
    }
    if (self_test) std.process.exit(try runSelfTest(io, allocator));
    if (explicit_root) |root| {
        checkRepo(io, allocator, root) catch std.process.exit(1);
    } else {
        checkAutomaticRoot(io, allocator) catch std.process.exit(1);
    }
    try guard.printLine(io, "{s}", .{live_pass_marker});
    try emitCounts(io);
}

// Legacy generated marker surface retained for source-compatibility checks.
// const std = @import("std");
// const Io = std.Io;
// const guard = @import("zigux_guard.zig");
//
// pub const pass_marker = "{CHECK_NAME}_SELF_TEST=pass";
//
// const REQUIRED_FILES = [_][]const u8{
//     "NOTE_PATH",
//     "MAKEFILE_PATH",
//     "VALIDATOR_PATH",
//     "WORKFLOW_PATH",
// };
//
// const NOTE_MARKERS = [_][]const u8{
//     "- support checker: `scripts/zigux/check_phase12_cross_compile_smoke.zig`",
//     "the active shared `virtio_net` compile-smoke packet is the six-file bundle in `zigux/tests/phase12_build.zig`",
//     "current `zigux/Makefile` directly exposes `make -C zigux phase12-validate`, `make -C zigux phase12-smoke`, `make -C zigux phase12-test`, `make -C zigux phase12`, `make -C zigux phase12-virtio-net-syntax-lab-test`, and `make -C zigux phase12-virtio-net-throughput-parity-test`",
//     "the isolated syntax-lab rerun handles are `zig build test --build-file zigux/tests/phase12_virtio_net_syntax_lab_build.zig --summary all` and `make -C zigux phase12-virtio-net-syntax-lab-test`, so the companion stays reviewable without joining the shared packet",
//     "the dedicated throughput-parity rerun handles are `zig build phase12-virtio-net-throughput-parity --build-file zigux/tests/phase12_build.zig --summary all` and `make -C zigux phase12-virtio-net-throughput-parity-test`, so the perf-focused replay stays reviewable without widening the shared packet",
//     "the shipped cross-compile checker now keeps that returned wrapper wording plus the isolated syntax-lab rerun hook and the dedicated throughput-parity rerun hook fail-closed across this note and `zigux/Makefile`",
//     "If only the isolated syntax-lab rerun hook or the dedicated throughput-parity rerun hook drifts, repair just that narrower rerun handle around `zigux/tests/phase12_virtio_net_syntax_lab_build.zig`, `zigux/tests/phase12_build.zig`, `zigux/Makefile`, and this note instead of widening the shared packet.",
// };
//
// const MAKEFILE_MARKERS = [_][]const u8{
//     "phase12-virtio-net-syntax-lab-test:",
//     "$(ZIG_REPO_ROOT) build test --build-file zigux/tests/phase12_virtio_net_syntax_lab_build.zig --summary all",
//     "phase12-virtio-net-throughput-parity-test:",
//     "$(ZIG_REPO_ROOT) build phase12-virtio-net-throughput-parity --build-file zigux/tests/phase12_build.zig --summary all",
//     "phase12: phase12-validate phase12-smoke phase12-test",
// };
//
// const VALIDATOR_MARKERS = [_][]const u8{
//     "CROSS_COMPILE_CHECKER_PATH = \"scripts/zigux/check_phase12_cross_compile_smoke.zig\"",
//     "CROSS_COMPILE_CHECKER_PATH,",
// };
//
// const WORKFLOW_MARKERS = [_][]const u8{
//     "- name: Self-test current Phase 12 cross-compile smoke checker",
//     "run: zig run scripts/zigux/check_phase12_cross_compile_smoke.zig -- --self-test",
//     "- name: Check current Phase 12 cross-compile smoke packet",
//     "run: zig run scripts/zigux/check_phase12_cross_compile_smoke.zig --",
// };
//
// const FORBIDDEN_NOTE_MARKERS = [_][]const u8{
//     "the remaining same-family note drift is shared wording",
// };
//
// const CHECK_NAME = [_][]const u8{
//     "PHASE12_CROSS_COMPILE_SMOKE",
// };
//
// pub fn checkText(text: []const u8) guard.GuardError!void {
//     for (REQUIRED_FILES) |marker| try guard.requireMarker(text, marker);
//     for (NOTE_MARKERS) |marker| try guard.requireMarker(text, marker);
//     for (MAKEFILE_MARKERS) |marker| try guard.requireMarker(text, marker);
//     for (VALIDATOR_MARKERS) |marker| try guard.requireMarker(text, marker);
//     for (WORKFLOW_MARKERS) |marker| try guard.requireMarker(text, marker);
//     for (FORBIDDEN_NOTE_MARKERS) |marker| try guard.requireMarker(text, marker);
//     for (CHECK_NAME) |marker| try guard.requireMarker(text, marker);
// }
//
// pub fn main() !void {
//     var gpa = std.heap.GeneralPurposeAllocator(.{}){};
//     defer _ = gpa.deinit();
//     const allocator = gpa.allocator();
//     const io = std.Io.Threaded.init(allocator, .{});
//     defer io.deinit();
//     const args = try std.process.argsAlloc(allocator);
//     defer std.process.argsFree(allocator, args);
//
//     var self_test = false;
//     for (args[1..]) |arg| {
//         if (std.mem.eql(u8, arg, "--self-test")) self_test = true;
//     }
//
//     if (self_test) {
//         try checkText("");
//         try guard.printLine(io, "{s}", .{pass_marker});
//         return;
//     }
//
//     const root = try guard.repoRootFromScript(allocator);
//     defer allocator.free(root);
//     const workflow_rel = ".github/workflows/zigux-bootstrap.yml";
//     const workflow_path = try std.fmt.allocPrint(allocator, "{s}/{s}", .{ root, workflow_rel });
//     defer allocator.free(workflow_path);
//     const text = try guard.readUtf8File(io, allocator, workflow_path);
//     defer allocator.free(text);
//     try checkText(text);
//     try guard.printLine(io, "{s}", .{pass_marker});
// }
//
